src/parser/newast.c

```diff
@@ -72,6 +93,39 @@ static const cypher_astnode_t *_NEWAST_GetClause(const cypher_astnode_t *node, c
     return NULL;
 }
 
+/* Collect references to all MATCH clauses in the query. Since clauses cannot be nested, we only need to
+ * check the immediate children of the query node.
+ * TODO Migrate other validations to this collection logic to improve performance
+ * (keeping in mind that most clauses can only appear once, simplifying this further). */
+static unsigned int _NEWAST_GetMatchClauses(const cypher_astnode_t *query, unsigned int clause_count, const cypher_astnode_t **matches) {
+  unsigned int match_count = 0;
+    for (unsigned int i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *child = cypher_astnode_get_child(query, i);
+        if (cypher_astnode_type(child) == CYPHER_AST_MATCH) {
+            matches[match_count] = child;
```

Comment:
`matches` is a stack allocation of size of size `clause_count`. I set it up this way to avoid a short-lived heap allocation, but don't mind changing it so that this function computes clause count and mallocs.